packages/forms/src/__tests__/form.test.tsx

```diff
@@ -254,6 +254,64 @@ describe('Form', () => {
     )
   })
 
+  it('ensures required textField is enforced by validation', async () => {
+    const mockFn = jest.fn()
+
+    render(
+      <Form onSubmit={mockFn}>
+        <TextField
+          name="userId2"
+          defaultValue=""
+          validation={{ required: true }}
+        />
+        <Submit>Save</Submit>
+      </Form>
+    )
+
+    fireEvent.click(screen.getByText('Save'))
+
+    // The validation should catch and prevent the onSubmit from being called
+    await waitFor(async () => {
+      await new Promise((res) =>
+        setTimeout(() => {
```

Comment:
Is this timeout really needed here?
At least testing the inverse (that a function has been called) can/should be done without it
https://testing-library.com/docs/dom-testing-library/api-events/#using-jest-function-mocks

Reply:
I've found it to be required.  If we put something like:
`    expect(mockFn).not.toHaveBeenCalled()`
on a function that is actually called, it will typically pass because it doesn't wait and see if a call will eventually be made.  If we put something like: 
`    await waitFor(() => expect(mockFn).not.toHaveBeenCalled())`
We get a similar passing response ... it passes before waiting for the function.  If know of a more concise way rather than using the promise and timeout, let me know.

Reply:
Why does it work for the example testing-library have in their docs and not for our code? What's different?

Reply:
Are there any examples that are using `expect(mockFn).not.toHaveBeenCalled()`?  I didn't see any referenced from 
[(https://testing-library.com/docs/dom-testing-library/api-events/#using-jest-function-mocks)](https://testing-library.com/docs/dom-testing-library/api-events/#using-jest-function-mocks) but perhaps I missed it.

Feel free to try it for yourself ... 
`expect(mockFn).toHaveBeenCalled() `will work reliably.
`expect(mockFn).not.toHaveBeenCalled()` will often pass regardless of whether the mockFn is called or not.

Here, I've tweaked one of the tests in our code .... the below will pass the test:
```
  it('sets the value to undefined for empty string on optional relational fields', async () => {
    const mockFn = jest.fn()

    render(
      <Form onSubmit={mockFn}>
        {/* This is an optional relational field because the name ends with "Id"
            and it doesn't have { required: true } */}
        <TextField name="userId" defaultValue="" />
        <SelectField name="groupId" defaultValue="">
          <option value="">No group</option>
          <option value={1}>Group 1</option>
          <option value={2}>Group 2</option>
        </SelectField>
        <Submit>Save</Submit>
      </Form>
    )

    fireEvent.click(screen.getByText('Save'))
    expect(mockFn).not.toHaveBeenCalled()

    await waitFor(() => expect(mockFn).toHaveBeenCalledTimes(1))
    expect(mockFn).toBeCalledWith(
      { userId: undefined, groupId: undefined },
      expect.anything() // event that triggered the onSubmit call
    )
  })
```

So, we could omit the timeout ... but we'd be tricking ourselves.

Reply:
@Tobbe, if you are interested, we could set up a quick video call and I could walk through it with you.

Reply:
Try this instead for 'ensures required selectField is enforced by validation':

```js
await act(async () => {
  fireEvent.submit(screen.getByText('Save'))
})

// The validation should catch and prevent the onSubmit from being called
expect(mockFn).not.toHaveBeenCalled()
```

Reply:
Sweet!  Learned something new ... I've never used act before.
Thanks for the tip!

Reply:
Or even this:
```
fireEvent.submit(screen.getByText('Save'))
await act(async () => {})

// The validation should catch and prevent the onSubmit from being called
expect(mockFn).not.toHaveBeenCalled()
```

But I'm not loving that either. Because Kent himself says not to do that. I'm just not sure what to do instead. We really have nothing to `waitFor`...

https://twitter.com/kentcdodds/status/1330937801093697538

Reply:
A more idiomatic way to test is probably like this:
```js
  it('ensures required selectField is enforced by validation', async () => {
    const mockFn = jest.fn()

    render(
      <Form onSubmit={mockFn}>
        <SelectField
          name="groupId2"
          defaultValue=""
          validation={{ required: true }}
        >
          <option value="">No group</option>
          <option value={1}>Group 1</option>
          <option value={2}>Group 2</option>
        </SelectField>
        <FieldError name="groupId2" data-testid="err" />
        <Submit>Save</Submit>
      </Form>
    )

    fireEvent.submit(screen.getByText('Save'))

    await waitFor(() => screen.getByTestId('err'))

    // The validation should catch and prevent the onSubmit from being called
    expect(mockFn).not.toHaveBeenCalled()
  })
  ```
  
  Or, possibly, remove `data-testid` and look for the actual validation error instead:
  ```
  await waitFor(() => screen.getByText('groupId2 is required'))
  ```
  
  I actually think using `getByText` even more idiomatic. But I don't like it in this case because we would have to update this testcase if we decided to change the validation test, which really this testcase cares nothing about.

Reply:
We're on a similar page.... I just tested out using the FieldError to waitfor

```
it('for a TextAreaField with valueAsJSON.  Bad JSON case 3', async () => {
    const mockFn = jest.fn()

    render(
      <Form onSubmit={mockFn}>
        <TextAreaField
          name="jsonField"
          defaultValue="{bad-json}"
          data-testid="jsonField"
          validation={{ valueAsJSON: true }}
        />
        <FieldError name="jsonField" data-testid="FieldError" />
        <Submit>Save</Submit>
      </Form>
    )
    //    await act(async () => {
    fireEvent.submit(screen.getByText('Save'))
    //    })
    await waitFor(() =>
      expect(screen.getByTestId('FieldError')).toBeInTheDocument()
    )

    // The validation should catch and prevent the onSubmit from being called
    expect(mockFn).not.toHaveBeenCalled()
  })

```

Works good!

Reply:
I think I prefer the `data-testid`.

I'll move in this direction if you're ok with it.

Reply:
Sounds good. Do it! 🙂